Add support for 300 Multiple Choices response.

bambyster opened the pull request:
Here is a change needed to support multiple redirect destinations. The patch also contains missed import of CCEventPreConnect in UacStateTrying and support for q= parameter in SipAddress.

Please note that the CCEventRedirect and CCEventDisconnect now carry the redirect destinations as a list of SipAddress-es instead of a single SipURL.

The changed code successfully passes the voiptests.

sobomax commented on sippy/UacStateTrying.py:
```diff
@@ -28,7 +28,7 @@
 from UaStateGeneric import UaStateGeneric
 from Timeout import TimeoutAbs
-from CCEvents import CCEventRing, CCEventConnect, CCEventFail, CCEventRedirect, CCEventDisconnect
+from CCEvents import CCEventRing, CCEventConnect, CCEventFail, CCEventRedirect, CCEventDisconnect, CCEventPreConnect
```
This change has to be committed separately. It's unrelated bug.

sobomax commented on sippy/SipAddress.py:
```diff
@@ -163,3 +164,11 @@ def setParam(self, name, value = None):
+
+    def getQ(self):
+        if self.q == None:
+            try:
```
Is there any reason to hide error and fake q value if some garbage is received? None of the other SIP headers have any try/except logic and there is a good reason to not do it.

sobomax replied:
Also, by making copy into self.q and not using it in the str() / localStr() methods is a good way to introduce a bug. Since setting x.q = 100.0 won't have any effect. I'd suggest something like this instead:

```python
                k_v = l.split('=', 1)
                if len(k_v) == 2:
                    k, v = k_v
                else:
                    k = k_v[0]
                    v = None
                if k == 'q':
                    v = float(v)
```
You would also need proper `'%.1f' % v` in the localStr().

sobomax commented on sippy/UaStateConnected.py:
```diff
@@ -151,13 +151,13 @@ def recvEvent(self, event):
                 req = self.ua.genRequest('REFER', reason = event.reason)
                 self.ua.lCSeq += 1
-                also = SipReferTo(address = SipAddress(url = redirect))
+                also = SipReferTo(redirects[0])
```
This is incorrect. When multiple redirects are provided, one with a highest value needs to be selected, not the first one. For that reason I suggested a method to be implemented in the CCEventXX to retrieve redirects in the proper order. Alternatively, sorting can be done during the generation of the event to avoid hitting parsing issues when originating outbound call.

sobomax replied:
That all being said, there is no way we can logically get more than one redirect in the CCEventDisconnect (see above), so that the whole code can be removed.

sobomax commented on sippy/UaStateConnected.py:
```diff
@@ -56,8 +56,8 @@ def recvRequest(self, req):
-            also = req.getHFBody('refer-to').getUrl().getCopy()
-            self.ua.equeue.append(CCEventDisconnect(also, rtime = req.rtime, origin = self.ua.origin))
+            also = req.getHFBody('refer-to').getCopy()
+            self.ua.equeue.append(CCEventDisconnect([ also ], rtime = req.rtime, origin = self.ua.origin))
```
Is there any reason to allow multiple contacts in the CCEventDisconnect? I see none really.

sobomax commented on sippy/UacStateRinging.py:
```diff
@@ -130,7 +130,10 @@ def recvResponse(self, resp, tr):
             return rval
         if code in (301, 302) and resp.countHFs('contact') > 0:
-            scode = (code, reason, body, resp.getHFBody('contact').getUrl().getCopy())
+            scode = (code, reason, body, [ resp.getHFBody('contact').getCopy() ])
```
I'd suggest using a list (immutable), not a tuple here.

bambyster commented:
I've reverted the changes to the CCEventDisconnect class and used tuples instead of lists.
Also I removed the helper code for the q= parameter from the SipAddress class and did not add any sorting for the redirect items. The rationale behind this is that the ordering by the q parameter is not declared as mandatory. The excerpts from RFC-3261:
**10.2.1.2 Preferences among Contact Addresses**
   _This list can be prioritized with the "q" parameter in the Contact header field._
**16.6 Request Forwarding**
   _A common ordering mechanism is to use the qvalue parameter of targets
   obtained from Contact header fields..._
If my understanding is correct, the RFC-3261 does not mandate the q parameter to be used as a sorting value but just states that the common practice is to use it for sorting. And so I can imagine that someone could want to sort the items differently or even use the q parameter as a filter. So IMO the best solution is to leave the sorting to the business logic of the software that uses sippy. The same idea applies to the UaStateConnected.py - if the user of the sippy library wants the REFER to be sent to the redirect address with the highest q value then he should intercept the CCEventRedirect, reorder the redirect list accordingly and pass the modified CCEventRedirect further.

sobomax commented on sippy/UasStateTrying.py:
```diff
@@ -79,14 +79,17 @@ def recvEvent(self, event):
                 return (UaStateConnected, self.ua.conn_cbs, event.rtime, event.origin)
             else:
-                self.ua.sendUasResponse(code, reason, body, self.ua.lContact, ack_wait = True, \
+                self.ua.sendUasResponse(code, reason, body, [ self.ua.lContact ], ack_wait = True, \
```
[ self.ua.lContact ] -> (self.ua.lContact,)

sobomax commented on sippy/UasStateTrying.py:
```diff
@@ -70,7 +70,7 @@ def recvEvent(self, event):
                 self.ua.no_progress_timer = None
             if isinstance(event, CCEventConnect):
-                self.ua.sendUasResponse(code, reason, body, self.ua.lContact, ack_wait = False, \
+                self.ua.sendUasResponse(code, reason, body, [ self.ua.lContact ], ack_wait = False, \
```
[ self.ua.lContact ] -> (self.ua.lContact,)

sobomax commented on sippy/UasStateTrying.py:
```diff
@@ -79,14 +79,17 @@ def recvEvent(self, event):
                 scode = (500, 'Failed', None, None)
-            self.ua.sendUasResponse(scode[0], scode[1], scode[2], SipContact(address = SipAddress(url = scode[3])))
+            elif scode[3] != None:
+                contacts = [ SipContact(address = x) for x in scode[3] ]
```
contacts = tuple(SipContact(address = x) for x in scode[3])

bambyster commented:
I've completed the change from list to tuple, thank you for noticing. And the CCEventDisconnect now contains a SipAddress not SipURL.
